### PR TITLE
Extend relationship with attributes between parent and child entity

### PR DESCRIPTION
**Pull request by martin-fleck-at**
- Add relationship attribute support in grammar
- Provide only correct attributes as suggestion to the user
- Validate to only have correct attributes and no duplicates

### CODE REVIEW

File: extensions/crossmodel-lang/src/language-server/relationship.langium [hunk 1]
```diff
@@ -4,20 +4,24 @@ import 'entity'
...
-        'child' ':' child=[Entity:IDReference] |
-        'type' ':' type=STRING
-    )
-;
\ No newline at end of file
+ 
+RelationshipAttribute:
+    '-' 'parent' ':' parent=[Attribute:IDReference]
+        'child' ':' child=[Attribute:IDReference]
+        ('description' ':' description=STRING)?
```
Comment (harmen-xb) [inline extensions/crossmodel-lang/src/language-server/relationship.langium:26]: This description is better to remove. This should never be used.

**Review by harmen-xb — COMMENTED** [review]
@martin-fleck-at The PR looks quite good. I added some inline comments.

As just discusses we both agree here the crossmodel grammer can be simplified where the arbitrary order of the properties won't be supported anymore.

File: extensions/crossmodel-lang/src/language-server/cross-model-validator.ts [hunk 2]
```diff
@@ -83,6 +86,37 @@ export class CrossModelValidator {
...
+      // we check that each attribute actually belongs to their respective entity (parent, child)
+      // and that each attribute is only used once
+      const usedParentAttributes: Attribute[] = [];
+      const usedChildAttributes: Attribute[] = [];
+      for (const attribute of relationship.attributes) {
+         if (attribute.parent.ref) {
+            if (attribute.parent?.ref?.$container !== relationship.parent?.ref) {
+               accept('error', 'Not a valid parent attribute.', { node: attribute, property: 'parent' });
+            }
+            if (usedParentAttributes.includes(attribute.parent.ref)) {
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply extensions/crossmodel-lang/src/language-server/cross-model-validator.ts]: Good point, I fixed it!

File: extensions/crossmodel-lang/src/language-server/cross-model-validator.ts [hunk 3]
```diff
@@ -83,6 +86,37 @@ export class CrossModelValidator {
...
+               accept('error', 'Each parent attribute can only be referenced once.', { node: attribute, property: 'parent' });
+            } else {
+               usedParentAttributes.push(attribute.parent.ref);
+            }
+         }
+         if (attribute.child.ref) {
+            if (attribute.child?.ref?.$container !== relationship.child?.ref) {
+               accept('error', 'Not a valid child attribute.', { node: attribute, property: 'child' });
+            }
+            if (usedChildAttributes.includes(attribute.child.ref)) {
```
[the remark this answers is not in the document]
Comment (martin-fleck-at) [reply extensions/crossmodel-lang/src/language-server/cross-model-validator.ts]: Good point, I fixed it!

File: extensions/crossmodel-lang/src/language-server/relationship.langium (hunk 1 above)
Comment (martin-fleck-at) [reply extensions/crossmodel-lang/src/language-server/relationship.langium]: I removed it, thank you!

File: extensions/crossmodel-lang/test/language-server/cross-model-lang-diagram.test.ts [hunk 4]
```diff
@@ -2,54 +2,31 @@
...
 describe('CrossModel language Diagram', () => {
    describe('Diagram without nodes and edges', () => {
       test('Simple file for diagram', async () => {
-         const document = diagram1;
-         const parsedDocument = await parseDocument(services, document);
-         const model = parsedDocument.parseResult.value as CrossModelRoot;
-
-         expect(model).toHaveProperty('systemDiagram');
-         expect(parsedDocument.parseResult.lexerErrors.length).toBe(0);
-         expect(parsedDocument.parseResult.parserErrors.length).toBe(0);
```
Comment (harmen-xb) [inline extensions/crossmodel-lang/test/language-server/cross-model-lang-diagram.test.ts:24]: @martin-fleck-at Are these expect parse (on lexer and parser) not relavant?

Comment (harmen-xb) [reply extensions/crossmodel-lang/test/language-server/cross-model-lang-diagram.test.ts:24]: Never mind, I see it's moved into the utils now.

**Review by harmen-xb — APPROVED** [review]
@martin-fleck-at All changes look good.

I will approve it.

**Pull request merged**